Add bosch_shc supporting Bosch Smart Home Controller

tschamm opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
New integration which introduces support for devices managed by the Bosch Smart Home Controller system. PR for home-assistant.io doc and home-assistant brands repos are opened.
This first PR provides the binary_sensor platform, in the case of Bosch SHC this are door/window contacts and smoke detectors.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

Integration does not support yaml configuration. Instead, it is configured via a config flow.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
Link to [feature request thread](https://community.home-assistant.io/t/bosch-smart-home/115864) on `community.home-assistant.io`. As soon as this PR is done, we can provide additional platform support for the Bosch SHC system. Already implemented and working platforms are:
* Alarm control panel
* Binary sensor, e.g. door/window contact, smoke detector
* Climate
* Cover, i.e. shutter control
* Sensor, e.g. thermostat, power/energy, battery
* Switch, e.g. smart plug, light control
We would like to see this as an out-of-the-box component provided by the home assistant system.

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12842

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

tschamm committed:
Add Bosch Smart Home Controller integration and binary_sensor platform

tschamm committed:
Fixed pylint warnings

tschamm committed:
Untested files added to .coveragerc

homeassistant commented:
Hi @tschamm, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#34063).

Once you do that we will be able to review and accept this pull request.

Thanks!

tschamm committed:
Skip already configured host.
Fixed connection class.
Share data via entry_id instead of CONF_NAME.
Removed commented out code.
Removed async_setup_platform method from binary_sensor.

tschamm committed:
Fixed pylint errors.

tschamm committed:
Upgrade to boschshcpy==0.1.3.
Improved docstrings.

tschamm committed:
Moved title to top level in strings.json

tschamm committed:
Improved polling start/stop.
Register service for trigger SHC scenarios.
Add service for smoke detector function check.
Follow style guideline.
Increase boschshcpy version to 0.1.4

tschamm commented:
This is a kind reminder.

tschamm committed:
Common methods for all shc devices are moved
to entity class.
Moved .translations to translations folder.

tschamm committed:
Started working on improvements suggested in PR

zimmermannb commented:
Are there any upgrades on this regarding the Review? Would be great to see this within the next version.....

SirHannesSie commented:
Yeah, I would also love if this was finally reviewed

alemp commented:
Release it, please.

MartinHjelmare commented:
What's the relationship between Bosch Smart Home and Home Connect?

https://github.com/home-assistant/core/pull/29214

tschamm commented:
> What's the relationship between Bosch Smart Home and Home Connect?
> 
> #29214

Home Connect provides control for BSH (Bosch Siemens Hausgeräte) home appliance devices, like smart dish washers, refrigerators etc. Bosch Smart Home provides monitoring/control of your home, i.e. devices like door/window sensors, cover/light controls, temperature controls. Please don't ask me why Bosch provides both. AFAIK, Home Connect devices can be integrated within Bosch Smart Home.

Maulwurf16 commented:
To add to @tschamm 's answer: Bosch itself provides two different Apps. One is to control the iOT Devices (Bosch Smart Home <- this pull request), whereas the other app controls the Household items (Bosch Siemens Hausgeräte: dishwashers, dryers, washing machine, etc). These two Device classes are separated completely. By looking at the Bosch Smart Home Homepage (https://www.bosch-smarthome.com/de/de/produkte) it becomes somewhat clear which products are the focus of this pull request.

MartinHjelmare commented on homeassistant/components/boschshc/__init__.py:
```diff
@@ -0,0 +1,136 @@
+    shc_info = session.information
+    if shc_info.version == "n/a":
+        _LOGGER.warning("Unable to connect to Bosch Smart Home Controller API")
+        return False
```
Is this a non recoverable error or just a temporary connection problem? If it's the latter we could raise `ConfigEntryNotReady`.

MartinHjelmare commented on homeassistant/components/boschshc/__init__.py:
```diff
@@ -0,0 +1,136 @@
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, data[CONF_IP_ADDRESS])},
```
Ip addresses can change so we don't want to use that for ids. Mac address would be ok, or better serial number.

MartinHjelmare commented on homeassistant/components/boschshc/binary_sensor.py:
```diff
@@ -0,0 +1,58 @@
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the SHC binary sensor platform."""
+    device = []
```
Please rename this to entities.
```suggestion
    entities = []
```

MartinHjelmare commented on homeassistant/components/boschshc/binary_sensor.py:
```diff
@@ -0,0 +1,58 @@
+    session: SHCSession = hass.data[DOMAIN][config_entry.entry_id]
+
+    for binarysensor in session.device_helper.shutter_contacts:
+        device.append(
```
```suggestion
        entities.append(
```

MartinHjelmare commented on homeassistant/components/boschshc/binary_sensor.py:
```diff
@@ -0,0 +1,58 @@
+            )
+        )
+
+    if device:
+        async_add_entities(device)
```
```suggestion
    if entities:
        async_add_entities(entities)
```

MartinHjelmare commented on homeassistant/components/boschshc/binary_sensor.py:
```diff
@@ -0,0 +1,58 @@
+    @property
+    def is_on(self):
+        """Return the state of the sensor."""
+        if self._device.state == SHCShutterContact.ShutterContactService.State.OPEN:
+            return True
+        if self._device.state == SHCShutterContact.ShutterContactService.State.CLOSED:
+            return False
+        return None
```
We can simplify this.
```suggestion
        return self._device.state == SHCShutterContact.ShutterContactService.State.OPEN
```

MartinHjelmare commented on homeassistant/components/boschshc/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+                info = await validate_input(self.hass, user_input)
+
+                # Check if already configured
+                await self.async_set_unique_id(user_input[CONF_IP_ADDRESS])
```
Ip address is not a valid unique id as it can change. The unique id must both be unique and stable.

If there's no valid unique id, we can just check if the ip address is in existing config entries for this domain and abort if so.

MartinHjelmare commented on homeassistant/components/boschshc/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+                self._abort_if_unique_id_configured()
+
+                return self.async_create_entry(title=info["title"], data=user_input)
+            except CannotConnect:
```
`CannotConnect` is never raised.

MartinHjelmare commented on homeassistant/components/boschshc/config_flow.py:
```diff
@@ -0,0 +1,90 @@
+        vol.Required(CONF_IP_ADDRESS): str,
+        vol.Required(CONF_SSL_CERTIFICATE): str,
+        vol.Required(CONF_SSL_KEY): str,
+        vol.Optional(CONF_ICON): str,
```
Icon is not used. Please remove it.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+            self.schedule_update_ha_state()
+
+        for service in self._device.device_services:
+            service.subscribe_callback(self.entity_id, on_state_changed)
```
The entity_id is not a stable id as the user can change it.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+        return self._device.serial
+
+    @property
+    def device_id(self):
```
We don't need this property. Just use `self._device.id`.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+        return self._device.id
+
+    @property
+    def root_device(self):
```
This isn't used.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+        return self._device.name
+
+    @property
+    def manufacturer(self):
```
Just use `self._device.manufacturer` directly.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+            "name": self.name,
+            "manufacturer": self.manufacturer,
+            "model": self._device.device_model,
+            "sw_version": "",
```
Don't set `sw_version` if we don't know it.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+    @property
+    def available(self):
+        """Return false if status is unavailable."""
+        if self._device.status == "AVAILABLE":
+            return True
+        return False
```
```suggestion
        return self._device.status == "AVAILABLE"
```

MartinHjelmare replied:
Consider breaking out magic strings to constants.

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+        self._device.update()
+
+    @property
+    def state_attributes(self):
+        """Extend state attribute of the device."""
+        state_attr = super().state_attributes
+        if state_attr is None:
+            state_attr = dict()
```
```suggestion
    def device_state_attributes(self):
        """Extend state attribute of the device."""
        state_attr = {}
```

MartinHjelmare commented on homeassistant/components/boschshc/entity.py:
```diff
@@ -0,0 +1,97 @@
+        """Report polling mode. SHC Entity is communicating via long polling."""
+        return False
+
+    def update(self):
```
`update` will never be called since this entity is not polling.

tschamm replied:
In an inherited class, `should_poll` may be overloaded, thus `update` is useful. Thus I'ld like to stick with it.

MartinHjelmare commented on homeassistant/components/boschshc/manifest.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "domain": "boschshc",
```
We prefer to separate words with underscore.
```suggestion
  "domain": "bosch_shc",
```

MartinHjelmare commented on homeassistant/components/boschshc/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "domain": "boschshc",
+  "name": "Bosch Smart Home Controller Interface API",
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/boschshc",
```
```suggestion
  "documentation": "https://www.home-assistant.io/integrations/bosch_shc",
```

MartinHjelmare commented on homeassistant/components/boschshc/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/boschshc",
+  "requirements": ["boschshcpy==0.1.6"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
Please remove empty strings.
```suggestion
```

MartinHjelmare commented on homeassistant/components/boschshc/strings.json:
```diff
@@ -0,0 +1,25 @@
+        "title": "SHC access parameters",
+        "data": {
+          "name": "Name",
+          "ip_address": "IP Address",
```
Please use as many common strings as possible.

Here are the common strings:
https://github.com/home-assistant/core/blob/4170eb0f83ee5fef11eef1b656d3aa4696c67b95/homeassistant/strings.json#L20-L55

Here is an example:
https://github.com/home-assistant/core/blob/4170eb0f83ee5fef11eef1b656d3aa4696c67b95/homeassistant/components/adguard/strings.json#L7

MartinHjelmare reviewed (changes requested):
The config flow requires tests and 100% test coverage. The scaffold script will help you get started on the tests.

tschamm commented:
I've started working on the requested changes. The config flow requires tests and 100% test coverage is still open. I'll leave another comment as soon as I'm ready for next review iteration.

tschamm committed:
Update __init__.py

Removed unneeded line

lunetics commented:
:+1:

christianreiss commented:
We're all eagerly awaiting the merge. Any news?

tschamm committed:
Adding config_flow test for bosch_shc component.

tschamm committed:
Updating to boschshcpy==0.1.10

tschamm committed:
Fixed config_flow test for bosch_shc.

tschamm committed:
Switch from ip_address to host.
Remove setup via yaml config.

tschamm committed:
Use getmac to obtain mac address for unique id of SHC.

tschamm committed:
Fixed config_flow test for Bosch SHC

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -59,7 +62,17 @@ async def async_step_user(self, user_input=None):
                 self._abort_if_unique_id_configured()
 
-                return self.async_create_entry(title=info["title"], data=user_input)
+                mac_address = await self.async_get_mac(user_input[CONF_HOST])
```
We prefer not to use getmac to get the mac address from the ip address since it doesn't work in all network environments.

colormodzz commented:
Hi

tschamm committed:
Revert using getmac in component. Mac address is provided by boschshcpy.

tschamm committed:
Additional config_flow tests for bosch_shc.

MartinHjelmare commented:
Please exclude the modules explicitly from `.coveragerc` that we don't want to calculate coverage for. Ie all except the `config_flow.py` module.

tschamm committed:
Excluded modules from .coveragerc.
Removed unnecessary async_step_import.

tschamm commented on homeassistant/components/boschshc/__init__.py:
```diff
@@ -0,0 +1,136 @@
+
+ENTITY_ID_FORMAT = DOMAIN + ".{}"
+
+CONFIG_SCHEMA = vol.Schema(
```
Following ADR-0010, I removed the `async_step_import` from `config_flow.py`. If I understood correctly, this invalidates the above mentioned **Example entry for configuration.yaml**. 
Shall I remove the example entry from the description above?

MartinHjelmare replied:
Yes, please.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,130 @@
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.bosch_shc.config_flow.validate_input",
```
Patch the library instead of our code in the config flow. That will make the coverage check pass.

tschamm committed:
Fixed config_flow tests for bosch_shc.

tschamm committed:
Updated strings.json and translations

tschamm commented:
Thanks again MartinHjelmare for your thorough review and your supporting comments. I have now figured out (better) how to write the tests.

tschamm committed:
Execute shc info acquisition in async job.

tschamm committed:
Handle polling not yet initialized when unloading config entry

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,164 @@
+    )
+
+    session_information = session.information
+    assert session_information is None
```
This test doesn't test the config flow. It just tests the library session. We don't need to do that.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,164 @@
+    )
+
+    with patch(
+        "homeassistant.components.bosch_shc.config_flow.validate_input",
```
Don't patch our code. Patch the library instance.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,164 @@
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    result2 = await hass.config_entries.flow.async_configure(
```
What are we testing here? There's no patch applied.

MartinHjelmare commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,119 @@
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Bosch SHC component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
```
There's no config.

MartinHjelmare commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,119 @@
+    device_registry = await dr.async_get_registry(hass)
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, dr.format_mac(shc_info.macAddress))},
```
If the mac address is None this id isn't unique.

MartinHjelmare commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,119 @@
+            EVENT_HOMEASSISTANT_STOP, stop_polling
+        )
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, start_polling)
```
Why do we wait until home assistant start to start polling?

MartinHjelmare commented on homeassistant/components/bosch_shc/binary_sensor.py:
```diff
@@ -0,0 +1,54 @@
+            ShutterContactSensor(
+                device=binarysensor,
+                room_name=session.room(binarysensor.room_id).name,
+                shc_uid=format_mac(session.information.macAddress),
```
`format_mac` expects a string. It will error with `None`.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+                info = await validate_input(self.hass, user_input)
+
+                # Check if already configured
+                await self.async_set_unique_id(user_input[CONF_HOST])
```
We don't allow ip address as unique id.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,88 @@
+                        CONF_SSL_KEY: user_input[CONF_SSL_KEY],
+                    },
+                )
+            except CannotConnect:
```
`CannotConnect` is never raised.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,77 @@
+        """Extend state attribute of the device."""
+        state_attr = {}
+        if self._room_name:
+            state_attr["bosch_shc_room_name"] = self._room_name
```
Is the room name dynamic? We use the state attributes for dynamic info related to the state of the entity. Only exception is for static info that is essential for automations.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,77 @@
+        """Report polling mode. SHC Entity is communicating via long polling."""
+        return False
+
+    def update(self):
```
This is never used since this entity has `should_poll` set to False.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,77 @@
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
```
This logger is never used.

MartinHjelmare commented on homeassistant/components/bosch_shc/translations/de.json:
```diff
@@ -0,0 +1,24 @@
+{
```
Please remove the German translation.

MartinHjelmare commented on homeassistant/components/bosch_shc/strings.json:
```diff
@@ -0,0 +1,24 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "Device is already configured"
```
Please use as many common string references as possible.

tschamm committed:
Removed unnecessary calls while setting up integration.
Removed not used exception.
Adapted tests.
Removed static info from device_state_attributes.
Use common strings in strings.json

tschamm committed:
Use mac address as unique identifier.
Assure mac address is not None.

tschamm committed:
Fixed tests with mac address from boschshcpy

tschamm commented:
Thanks again for your review. The SHC device does not provide unique information except the mac address. I check for an invalid mac address now, is this sufficient? If not, I will skip registering the device, until the vendor (Bosch) provides the serial via their API.

tschamm committed:
Acquire unique info via mDNS

tschamm committed:
Reworked config_flow following shelly example. Flow is much more robust against errors.

tschamm committed:
Minor adaptions to config_flow. Adapted tests to new config_flow. Added tests for discovery.

tschamm committed:
Fixed flow_title, only show node_name in discovery. Updated manifest.

tschamm commented:
I reworked the external library to provide unique information on the bridge via mDNS. This helped me to better understand the  zeroconf discovery step.
I rewrote the config flow and the tests to support user configuration as well as zeroconf configuration. The integration now uses unique information provided from the external library.

tschamm committed:
Bumped to version 0.1.15

tschamm commented:
Hi, is there anything missing before a follow up review can take place?
Btw: thanks for the great work with HA 2020.12 and for setting up the HA conference.

MartinHjelmare commented:
This is on my todo list. I'm slowly getting there.

tschamm committed:
Bumped to boschshcpy==0.1.19
Updates from boschshc-hass custom component.

tschamm committed:
Fixed import of patch function

tschamm committed:
Merged in changes from HACS component. Bumped to boschshcpy==0.2.9

tschamm committed:
Bring test coverage back to 100%

tschamm commented:
Nearly celebrating this PR's first anniversary 🎂, I updated the code with code from `bosch_shc` custom component. I'm still looking forward to having this integration added to default components.

MartinHjelmare commented:
It's on my todo list.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+from homeassistant.components.zeroconf import async_get_instance
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_TOKEN
+
+from .const import (
+    CONF_HOSTNAME,
+    CONF_SHC_CERT,
+    CONF_SHC_KEY,
+    CONF_SSL_CERTIFICATE,
+    CONF_SSL_KEY,
+)
+from .const import DOMAIN  # pylint:disable=unused-import
```
We don't need this pylint disable anymore. It's been fixed upstream.
```suggestion
from .const import (
    CONF_HOSTNAME,
    CONF_SHC_CERT,
    CONF_SHC_KEY,
    CONF_SSL_CERTIFICATE,
    CONF_SSL_KEY,
    DOMAIN,
)
```

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+)
+
+
+async def validate_input(hass: core.HomeAssistant, host, cert, key):
```
Please type the functions completely or not at all.
```suggestion
async def validate_input(hass: core.HomeAssistant, host: str, cert: str, key: str) -> None:
```

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+    await hass.async_add_executor_job(session.authenticate)
+
+
+def write_tls_asset(hass: core.HomeAssistant, filename: str, asset: bytes):
```
```suggestion
def write_tls_asset(hass: core.HomeAssistant, filename: str, asset: bytes) -> None:
```

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                helper = await self.hass.async_add_executor_job(
+                    SHCRegisterClient, self.host, user_input[CONF_PASSWORD]
+                )
+                result = await self.hass.async_add_executor_job(
+                    helper.register, self.host, "HomeAssistant"
+                )
+                if "token" in result:
+                    self.hostname = result["token"].split(":", 1)[1]
+
+                await self.hass.async_add_executor_job(
+                    write_tls_asset, self.hass, CONF_SHC_CERT, result["cert"]
+                )
+                await self.hass.async_add_executor_job(
+                    write_tls_asset, self.hass, CONF_SHC_KEY, result["key"]
+                )
+                await validate_input(
+                    self.hass,
+                    self.host,
+                    self.hass.config.path(DOMAIN, CONF_SHC_CERT),
+                    self.hass.config.path(DOMAIN, CONF_SHC_KEY),
+                )
```
We should merge all six executor calls here into one call. Extract the sync calls into a function that we schedule once on the executor.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+                _LOGGER.warning(
+                    "SHC not in pairing mode! Please press the Bosch Smart Home Controller button until LED starts blinking"
+                )
+                errors["base"] = "unknown"
```
Why is this an unknown error? We seem to know what the problem is.

MartinHjelmare replied:
Bump.

tschamm replied:
I changed the behavior in the 3rd party library: wrong password or bridge not in pairing mode will both result in a registration error `SHCRegistrationError`, and are handled that way.
However, it can happen that session errors occur, when the API will return Non-OK result. This is why I treated them also as `unknown`

tschamm replied:
I introduced a new error type `session_error` for this type of error.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        except SHCConnectionError:
+            return self.async_abort(reason="cannot_connect")
+        except SHCmDNSError:
+            _LOGGER.exception("Error looking up mDNS entry")
```
This could become spammy for the user and the user would not have a good way to avoid it. I'd recommend logging at debug level instead and include the exception info in the log.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        await self.async_set_unique_id(info["unique_id"])
+        self._abort_if_unique_id_configured({CONF_HOST: zeroconf_info["host"]})
+        self.host = zeroconf_info["host"]
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
```
This is no longer needed.
```suggestion
```

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,220 @@
+        """Get additional information."""
+        zeroconf = await async_get_instance(self.hass)
+
+        session = await self.hass.async_add_executor_job(
+            SHCSession,
+            host,
+            "",
+            "",
+            True,
+            zeroconf,
+        )
+
+        information = await self.hass.async_add_executor_job(session.mdns_info)
```
Please merge the calls into a function that we schedule once on the executor.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+"""Bosch Smart Home Controller base entity."""
+from boschshcpy.device import SHCDevice
+
+from homeassistant.helpers.device_registry import async_get_registry as get_dev_reg
```
```suggestion
from homeassistant.helpers.device_registry import async_get as get_dev_reg
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+
+from homeassistant.helpers.device_registry import async_get_registry as get_dev_reg
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity_registry import async_get_registry as get_ent_reg
```
```suggestion
from homeassistant.helpers.entity_registry import async_get as get_ent_reg
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+
+async def get_device_id(hass, device_id):
+    """Get device id from device registry."""
+    dev_registry = await get_dev_reg(hass)
```
```suggestion
    dev_registry = get_dev_reg(hass)
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+from .const import DOMAIN
+
+
+async def get_device_id(hass, device_id):
```
This is only used once, it seems. We can inline it where it's needed.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+    await entity.async_remove()
+    ent_registry = await get_ent_reg(hass)
+    if entity.entity_id in ent_registry.entities:
+        ent_registry.async_remove(entity.entity_id)
```
If we remove the entity entry from the entity registry the `Entity` instance will be removed automatically.

tschamm replied:
Ok, this means I can remove the `entity.async_remove()` call?

MartinHjelmare replied:
I think so, yes.

tschamm replied:
I removed the `entity.async_remove()` call.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+    if entity.entity_id in ent_registry.entities:
+        ent_registry.async_remove(entity.entity_id)
+
+    dev_registry = await get_dev_reg(hass)
```
```suggestion
    dev_registry = get_dev_reg(hass)
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+async def remove_devices(hass, entity, entry_id):
+    """Get item that is removed from session."""
+    await entity.async_remove()
+    ent_registry = await get_ent_reg(hass)
```
```suggestion
    ent_registry = get_ent_reg(hass)
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+class SHCEntity(Entity):
+    """Representation of a SHC base entity."""
+
+    def __init__(self, device: SHCDevice, parent_id: str, entry_id: str):
```
```suggestion
    def __init__(self, device: SHCDevice, parent_id: str, entry_id: str) -> None:
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+
+        def update_entity_information():
+            if self._device.deleted:
+                self.hass.async_create_task(
```
The async api needs to be called from async context. We could use `hass.add_job` instead.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+                self.schedule_update_ha_state()
+
+        for service in self._device.device_services:
+            service.subscribe_callback(self.entity_id, on_state_changed)
```
Is this async safe?

tschamm replied:
`subscribe_callback` is an assignement in a dict, so yes.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+        """Name of the entity."""
+        return self.device_name
+
+    @property
+    def device_name(self):
+        """Name of the device."""
+        return self._device.name
+
+    @property
+    def device_id(self):
+        """Return the ID of the device."""
+        return self._device.id
+
```
We can remove these properties and use the instance attributes directly.
```suggestion
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+        """Return the device info."""
+        return {
+            "identifiers": {(DOMAIN, self._device.id)},
+            "name": self.device_name,
```
```suggestion
            "name": self._device.name,
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+    @property
+    def name(self):
+        """Name of the entity."""
+        return self.device_name
```
```suggestion
        return self._device.name
```

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,112 @@
+    def device_info(self):
+        """Return the device info."""
+        return {
+            "identifiers": {(DOMAIN, self._device.id)},
```
What's the difference between the device serial and the device id?

tschamm replied:
device id is a string like `hdm:HomeMaticIP:0123...`, while serial is just the serial number `0123...`. Both are AFAIK unique.

MartinHjelmare replied:
Ok.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,621 @@
+            {"host": "1.1.1.1"},
+        )
+
+    assert result2["type"] == "form"
```
Assert the step id too.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,621 @@
+    ), patch("homeassistant.components.bosch_shc.config_flow.write_tls_asset",), patch(
+        "boschshcpy.session.SHCSession.authenticate",
+    ), patch(
+        "homeassistant.components.bosch_shc.async_setup", return_value=True
+    ) as mock_setup, patch(
```
```suggestion
```

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,621 @@
+        },
+    ), patch("homeassistant.components.bosch_shc.config_flow.write_tls_asset",), patch(
+        "boschshcpy.session.SHCSession.authenticate"
+    ) as mock_setup, patch(
```
```suggestion
    ) as mock_authenticate, patch(
```

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,621 @@
+        "hostname": "123",
+    }
+
+    assert len(mock_setup.mock_calls) == 1
```
```suggestion
    assert len(mock_authenticate.mock_calls) == 1
```

tschamm committed:
Changes proposed from review.

tschamm committed:
Added iot_class to manifest.

tschamm commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,108 @@
+            data[CONF_SSL_CERTIFICATE],
+            data[CONF_SSL_KEY],
+            False,
+            zeroconf,
```
Zeroconf instance is passed to the library to obtain the unique id (unique name, mac address) via mDNS.

MartinHjelmare replied:
Ok, that's ok. Note that it won't work if the device is on a different subnet/lan.

tschamm committed:
Added reauth step if authentication fails

tschamm commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,108 @@
+        )
+    except SHCAuthenticationError as err:
+        _LOGGER.warning("Unable to authenticate on Bosch Smart Home Controller API")
+        raise ConfigEntryNotReady from err
```
Makes absolute sense, setup entry will return False then. I will add reauth later.

tschamm commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,108 @@
+        await hass.async_add_executor_job(session.stop_polling)
+
+    await hass.async_add_executor_job(session.start_polling)
+    session.reset_connection_listener = hass.bus.async_listen_once(
```
No, `session.reset_connection_listener()` is only called within `async_unload_entry`.

MartinHjelmare replied:
I'd then not save this callback on the session since the session instance is part of the library.

tschamm committed:
Merge executor calls into one call.
Remove mDNS exception handling.
Bumped to boschshcpy==0.2.15

tschamm committed:
Merge calls into a function.
Report error when not in pairing mode.

tschamm committed:
Fix return value in config_flow.Remove properties not needed. Rename async functions.

tschamm committed:
Make reauth working

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -1,32 +1,31 @@
         ent_registry.async_remove(entity.entity_id)
 
-    dev_registry = await get_dev_reg(hass)
+    async def get_device_id(hass, device_id):
```
We don't need a function for this if it's only used here.

tschamm committed:
Save callback outside of session

tschamm committed:
Test step ids and reauth.
Cleaned up tests.

tschamm committed:
Remove child class implementation of async_create_entry.

tschamm committed:
Use correct iot class.

tschamm committed:
Revert removing device_id from entity.
Fixed pylint errors.

tschamm committed:
Renamed polling handler

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,228 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                zeroconf = await async_get_instance(self.hass)
```
Please move lines that can't raise out of the try... except block.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,228 @@
+                    user_input,
+                    zeroconf,
+                )
+                entry_data = {
```
Move this down.

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,228 @@
+
+        schema = vol.Schema(
+            {
+                vol.Required(CONF_PASSWORD, default=user_input.get(CONF_PASSWORD)): str,
```
```suggestion
                vol.Required(
                    CONF_PASSWORD, default=user_input.get(CONF_PASSWORD, "")
                ): str,
```

MartinHjelmare commented on homeassistant/components/bosch_shc/config_flow.py:
```diff
@@ -0,0 +1,228 @@
+                errors["base"] = "unknown"
+            except SHCRegistrationError:
+                _LOGGER.warning(
+                    "SHC not in pairing mode! Please press the Bosch Smart Home Controller button until LED starts blinking"
```
Please break long strings around 88 characters. Maybe this message should be included in the error message above the form?

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,105 @@
+    if entity.entity_id in ent_registry.entities:
+        ent_registry.async_remove(entity.entity_id)
+
+    async def async_get_device_id(hass, device_id):
```
This doesn't need to be a function. We only use it in one place.

MartinHjelmare commented on homeassistant/components/bosch_shc/entity.py:
```diff
@@ -0,0 +1,105 @@
+    dev_registry = get_dev_reg(hass)
+    device_id = async_get_device_id(hass, entity.device_id)
+    if device_id is not None:
+        dev_registry.async_update_device(device_id, remove_config_entry_id=entry_id)
```
Why don't we just remove the device entry corresponding to the removed bosch device from the device registry? All connected entity entries and entities will be removed automatically too.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,578 @@
+        assert result2["errors"] == {}
+
+    with patch(
+        "homeassistant.components.bosch_shc.config_flow.create_credentials_and_validate",
```
Please don't patch our code in the config flow. Patch in the periphery, eg the library, to avoid I/O.

tschamm replied:
I already tested the code in another test, so I patched this to prevent duplicate code. I guess your remark is valid for all patches like "homeassistant.components.bosch_shc.config_flow..."?

MartinHjelmare replied:
Yes, please always patch in the periphery. Sometimes it's ok to patch a different part of the integration but we don't want to patch the module we're aiming at testing.

MartinHjelmare commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,114 @@
+from homeassistant.config_entries import SOURCE_REAUTH, ConfigEntry
+from homeassistant.const import CONF_HOST, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
```
```suggestion
from homeassistant.exceptions import ConfigEntryAuthFailed, ConfigEntryNotReady
```

tschamm committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

tschamm commented on homeassistant/components/bosch_shc/__init__.py:
```diff
@@ -0,0 +1,114 @@
+    device_registry.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        connections={(dr.CONNECTION_NETWORK_MAC, dr.format_mac(shc_info.unique_id))},
+        identifiers={(DOMAIN, shc_info.name)},
```
I changed it to use unique_id instead, which is the mac address reported via the API.

MartinHjelmare replied:
If we use the mac address I don't think we need to specify `identifiers` since we use the same id in `connections`.

tschamm replied:
The "via_device" attribute of the entities should be the same as the identifier of the bridge, isn't it? Or can we change via_device to use the connection information?

MartinHjelmare replied:
Ah, yes, you're right. We need to keep it for that to work.

tschamm committed:
Apply comments from code review.

MartinHjelmare commented:
Please solve the merge conflict.

tschamm committed:
Patch periphery code instead of own code.

tschamm committed:
Bumped to boschshcpy==0.2.16.
Improved verbosity when an error occurs.

tschamm committed:
Bumped to boschshcpy==0.2.17.
Fixed tests.

MartinHjelmare commented on homeassistant/components/bosch_shc/strings.json:
```diff
@@ -0,0 +1,37 @@
+    "error": {
+      "cannot_connect": "[%key:common::config_flow::error::cannot_connect%]",
+      "invalid_auth": "[%key:common::config_flow::error::invalid_auth%]",
+      "pairing_failed": "Pairing failed... Please press the Bosch Smart Home Controller button until LED starts blinking and check your password.",
```
Please don't use ellipsis. We should be clear and concise in the error message. Use a period or semi-colon.

tschamm committed:
Improved error handling based on code review comments.

MartinHjelmare commented on tests/components/bosch_shc/test_config_flow.py:
```diff
@@ -0,0 +1,625 @@
+
+    assert result3["type"] == "form"
+    assert result3["step_id"] == "credentials"
+    assert result3["errors"] == {"base": "unknown"}
```
```suggestion
    assert result3["errors"] == {"base": "session_error"}
```

MartinHjelmare committed:
Fix test

MartinHjelmare reviewed (approved):
Great!

danielschmitz87 commented:
Is there a plan to add more devices like the thermostates, shutter, Smoke Detectors and Twinguards?

tschamm commented:
@danielschmitz87 Yes, we create PRs for each platform step by step, following the rules of the HA developers. To keep you updated, have a look at https://community.home-assistant.io/t/bosch-smart-home/115864/238
